### DISCUSSION TITLE
Duplicate entries in Wikidata widget

### DISCUSSION THREAD

**Discussion by EricZhang456** [searxng/searxng#1347] [Q&A]
The Wikidata widget is really useful, but sometimes it messes up with duplicate entries. Here is one for "Official Website", each one of them points to the same website, but in a different language.

![IMG_20220618_210645.jpg](https://user-images.githubusercontent.com/32814750/174439154-ddb60e52-8bc2-4a08-ab49-07bdce9cf167.jpg)

I know this is more of a Wikidata's fault but it would be nice of there are at least *some* ways to get around this.

Language is set to en_US in SearXNG's settings

**Comment by EricZhang456** [comment]
Just checked Wikidata. Each of these links are in fact marked as different languages

![IMG_20220618_211802.jpg](https://user-images.githubusercontent.com/32814750/174439472-148f681b-ad2d-4b0f-aed0-c921b6eebd55.jpg)

**Reply by return42** [reply]
Even it mess up, this is what is registered at wikidata .. open https://www.wikidata.org/wiki/Q3094963 and scroll down to section `official website` and look at the different values of the property `language of work or name` .. all the languages are registered to different URLs.

Do we often see such messing entries? .. I tried to search about other companies but all of them have only one *"Official Website"* registered.

**Reply by dalf** [reply]
At first sight, I'm not sure this is doable because it will increase the response time of the Wikidata engine.

@return42 > from the user point of view, if the requested language is German, then the user can expect to get a link to the German website not to other languages (especially without labels). This is what happen for the Wikipedia engine for example.

We can imagine this logic:
* if the language is not specified, then the website is displayed.
* if the language is specified, pick only the one that match the user preference, or fallback to the English one.

---

Technical notes for later investigation: 
* first method: a subclass of [WDURLAttribute](https://github.com/searxng/searxng/blob/8f7cb4ad6fa3f9dec4b89807a32fa69e0e72bb84/searx/engines/wikidata.py#L570-L590) should be created for official website. The `get_where` method (inherited from `WDAttribute`) should be replaced by query similar to https://w.wiki/5Jma

* second method: concat the value + language in SPARQL, then split the values in Python:
  * https://github.com/searxng/searxng/blob/8f7cb4ad6fa3f9dec4b89807a32fa69e0e72bb84/searx/engines/wikidata.py#L480
  * https://github.com/searxng/searxng/blob/8f7cb4ad6fa3f9dec4b89807a32fa69e0e72bb84/searx/engines/wikidata.py#L584
